holmes/plugins/toolsets/prometheus.py

```diff
@@ -0,0 +1,456 @@
+import os
+import re
+import logging
+import random
+import string
+import time
+
+from typing import Any, Union, Optional
+
+import requests
+from pydantic import BaseModel
+from holmes.core.tools import (
+    CallablePrerequisite,
+    Tool,
+    ToolParameter,
+    Toolset,
+    ToolsetTag,
+)
+import json
+from requests import RequestException
+
+from urllib.parse import urljoin
+
+from holmes.utils.cache import TTLCache
+
+
+class PrometheusConfig(BaseModel):
+    prometheus_url: Union[str, None]
+    # Setting to None will remove the time window from the request for labels
+    metrics_labels_time_window_hrs: Union[int, None] = 48
+    # Setting to None will disable the cache
+    metrics_labels_cache_duration_hrs: Union[int, None] = 12
+
+
+class BasePrometheusTool(Tool):
+    toolset: "PrometheusToolset"
+
+
+def generate_random_key():
+    return "".join(random.choices(string.ascii_letters + string.digits, k=4))
+
+
+def filter_metrics_by_type(metrics: dict, expected_type: str):
+    return {
+        metric_name: metric_data
+        for metric_name, metric_data in metrics.items()
+        if metric_data.get("type") == expected_type
+    }
+
+
+def filter_metrics_by_name(metrics: dict, pattern: str) -> dict:
+    regex = re.compile(pattern)
+    return {
+        metric_name: metric_data
+        for metric_name, metric_data in metrics.items()
+        if regex.search(metric_name)
+    }
+
+
+def fetch_metadata(url: str) -> dict:
+    metadata_url = urljoin(url, "/api/v1/metadata")
+    metadata_response = requests.get(metadata_url, timeout=60, verify=True)
+
+    metadata_response.raise_for_status()
+
+    metadata = metadata_response.json()["data"]
+    return metadata
+
+
+def result_has_data(result: dict) -> bool:
+    data = result.get("data", {})
+    if len(data.get("result", [])) > 0:
+        return True
+    return False
+
+
+def fetch_metrics_labels(
+    prometheus_url: str,
+    cache: Optional[TTLCache],
+    metrics_labels_time_window_hrs: Union[int, None],
+    metric_name: str,
+) -> dict:
+    """This is a slow query. Takes 5+ seconds to run"""
+    cache_key = f"metrics_labels:{metric_name}"
+    if cache:
+        cached_result = cache.get(cache_key)
+        if cached_result:
+            logging.info("fetch_metrics_labels() result retrieved from cache")
```

Comment:
is this really info?

Reply:
I would think so. What I'm thinking is that we do log the tool calls and such message could help users quickly understand if Holmes is missing some new metrics that were added recently.
I can move it to debug if you prefer.

Reply:
got it, thanks